### PR TITLE
networkd: Allow kernel to handle IPv6 RA

### LINKED ISSUE #9582: Allow ipv6 RA but without ip autoconf
Hi,

The configuration with my provider is quite strange. I have to activate ipv6 RA to obtain a default route. But I don't want to have the address via autoconf. For that, I use a dhcpv6 pd, with static ip configuration
after that.

It should be great to be able to keep RA active but without autoconf. Currently, in my .network
conffile, I've got this :
```
IPv6AcceptRA=True
```
and I've created a file /etc/sysctl.d/local-ipv6.conf :
```
  net.ipv6.conf.all.use_tempaddr = 0
  net.ipv6.conf.default.use_tempaddr = 0
  net.ipv6.conf.ens3.use_tempaddr = 0
  net.ipv6.conf.all.autoconf = 0
  net.ipv6.conf.default.autoconf = 0
  net.ipv6.conf.ens3.autoconf = 0
  net.ipv6.conf.all.accept_ra_pinfo = 0
  net.ipv6.conf.default.accept_ra_pinfo = 0
  net.ipv6.conf.ens3.accept_ra_pinfo = 0
```
but I still get an address automatically :
```
  inet6 xxx::xxx/64 scope global mngtmpaddr noprefixroute dynamic
```

Networkd should add an "IPv6Autoconf" configuration parameter to disable the configuration of ipv6 addresses via Router Advertisement. Maybe with 3 values : yes, no and kernel ?

### PR DESCRIPTION
**Pull request by ssahani**
Closes #9582

### CODE REVIEW

File: src/network/networkd-network.c [hunk 1]
```diff
@@ -929,6 +929,62 @@ int config_parse_ipv6_privacy_extensions(
         return 0;
 }
 
+static const char* const ipv6_accept_ra_table[_IPV6_ACCEPT_RA_MAX] = {
+        [IPV6_ACCEPT_RA_NO] = "no",
+        [IPV6_ACCEPT_RA_YES] = "yes",
+        [IPV6_ACCEPT_RA_KERNEL] = "kernel",
+};
+
+DEFINE_STRING_TABLE_LOOKUP(ipv6_accept_ra, IPv6AcceptRA);
```
Comment (yuwata) [inline src/network/networkd-network.c]: Please use `DEFINE_STRING_TABLE_LOOKUP_WITH_BOOLEAN()` macro. Then, `config_parse_ipv6_accept_ra()` can be defined through `DEFINE_CONFIG_PARSE_ENUM()` macro.

File: src/network/networkd-network.h [hunk 2]
```diff
@@ -49,6 +49,15 @@ typedef enum IPv6PrivacyExtensions {
         _IPV6_PRIVACY_EXTENSIONS_INVALID = -1,
 } IPv6PrivacyExtensions;
 
+typedef enum IPv6AcceptRA {
+        IPV6_ACCEPT_RA_NO,
+        IPV6_ACCEPT_RA_YES,
+        IPV6_ACCEPT_RA_KERNEL,
+        IPV6_ACCEPT_RA_KERNEL_DISABLED,
```
Comment (yuwata) [inline src/network/networkd-network.h]: I cannot find any difference between `IPV6_ACCEPT_RA_YES` and `IPV6_ACCEPT_RA_KERNEL_DISABLED`. Maybe, we can drop `IPV6_ACCEPT_RA_KERNEL_DISABLED`?

Also, please add brief comments about each values. I guess
`IPV6_ACCEPT_RA_NO`: disable RA on networkd and kernel.
`IPV6_ACCEPT_RA_YES`: enable RA on networkd, and disable it on kernel.
`IPV6_ACCEPT_RA_KERNEL`: disable RA on networkd, and enable it on kernel.
`IPV6_ACCEPT_RA_KERNEL_DISABLED`: what???

File: src/network/networkd-link.c [hunk 3]
```diff
@@ -2547,10 +2549,16 @@ static int link_set_ipv6_accept_ra(Link *link) {
...
-        /* We handle router advertisements ourselves, tell the kernel to GTFO */
-        r = write_string_file(p, "0", WRITE_STRING_FILE_VERIFY_ON_FAILURE | WRITE_STRING_FILE_DISABLE_BUFFER);
+        if (link->network->ipv6_accept_ra == IPV6_ACCEPT_RA_KERNEL)
+                k = 1;
+        else
+                k = 0;
+
+        xsprintf(buf, "%u", k);
+
+        r = write_string_file(p, buf, WRITE_STRING_FILE_VERIFY_ON_FAILURE | WRITE_STRING_FILE_DISABLE_BUFFER);
```
Comment (yuwata) [inline src/network/networkd-link.c]:
```
r = write_string_file(p, 
                     link->network->ipv6_accept_ra == IPV6_ACCEPT_RA_KERNEL ? "1" : "0", 
                     WRITE_STRING_FILE_VERIFY_ON_FAILURE | WRITE_STRING_FILE_DISABLE_BUFFER);
```
can shorten the code a bit.

**Comment by yuwata** [comment]
BTW, I am not sure this certainly closes #9582... Please add testcases if possible.

**Comment by ssahani** [comment]
> _IPV6_ACCEPT_RA_INVALID

why defaults is preserved as kernel RA is disabled.

**Comment by yuwata** [comment]
> > _IPV6_ACCEPT_RA_INVALID
> 
> why defaults is preserved as kernel RA is disabled.

Previously, if `IPv6AcceptRA=` is not set, then it is determined by ipv6_forward is enabled or not. But, with this, the condition `if (link->network->ipv6_accept_ra < 0)` in `link_ipv6_accept_ra_enabled()` never hits, as all possible assigned values are non-negative. Is it desired?

Also, if it is desired, why not to set the default value `IPV6_ACCEPT_RA_YES`? As I already commented, I cannot find any difference between `IPV6_ACCEPT_RA_YES` and `IPV6_ACCEPT_RA_KERNEL_DISABLED`...

**Comment by ssahani** [comment]
> > > _IPV6_ACCEPT_RA_INVALID
> > 
> > 
> > why defaults is preserved as kernel RA is disabled.
> 
> Previously, if `IPv6AcceptRA=` is not set, then it is determined by ipv6_forward is enabled or not. But, with this, the condition `if (link->network->ipv6_accept_ra < 0)` in `link_ipv6_accept_ra_enabled()` never hits, as all possible assigned values are non-negative. Is it desired?
> 
> Also, if it is desired, why not to set the default value `IPV6_ACCEPT_RA_YES`? As I already commented, I cannot find any difference between `IPV6_ACCEPT_RA_YES` and `IPV6_ACCEPT_RA_KERNEL_DISABLED`...

Since it was a tristate we were assigning -1 and now we have a enum we can define INVALID which is like networkd will handle the RA

**Comment by ssahani** [comment]
updated pushed

File: man/systemd.network.xml [hunk 4]
```diff
@@ -619,11 +619,12 @@
         </varlistentry>
         <varlistentry>
           <term><varname>IPv6AcceptRA=</varname></term>
-          <listitem><para>Takes a boolean. Controls IPv6 Router Advertisement (RA) reception support for the interface.
+          <listitem><para>Takes one of boolean or kernel. Controls IPv6 Router Advertisement (RA) reception support for the interface.
```
Comment (yuwata) [inline man/systemd.network.xml]: `Takes a boolean or the special value <literal>kernel</literal>.`

File: man/systemd.network.xml [hunk 5]
```diff
@@ -619,11 +619,12 @@
...
           If true, RAs are accepted; if false, RAs are ignored, independently of the local forwarding state.
           If unset, the kernel's default is used, and RAs are accepted only when local forwarding
           is disabled for that interface.  When RAs are accepted, they may trigger the start of the DHCPv6 client if
-          the relevant flags are set in the RA data, or if no routers are found on the link.</para>
+          the relevant flags are set in the RA data, or if no routers are found on the link. If set to kernel then
```
Comment (yuwata) [inline man/systemd.network.xml]: `If set to <literal>kernel</literal> then`

**Comment by yuwata** [comment]
@ssahani I've slightly updated your commit, and added a tiny test for `IPv6AcceptRA=kernel`.

**Comment by yuwata** [comment]
BTW, to support #9582 more friendly, and especially #10647, it may be good to add `AcceptPrefix=` or something, which takes `all` or a list of prefixes.

**Comment by ssahani** [comment]
> @ssahani I've slightly updated your commit, and added a tiny test for `IPv6AcceptRA=kernel`.

Ok. 

> AcceptPrefix= 
 
This is can be done only via networkd will be a separate patch

**Comment by yuwata** [comment]
> > AcceptPrefix=
> 
> This is can be done only via networkd will be a separate patch

Yeah, of course.

**Comment by poettering** [comment]
hmm, what's the usecase for this? can you elaborate on why the kernel RA implementation is preferable over our own? To me this appears to be a work-around, and we should rather fix our own RA implementation if it doesn't deliver what it should deliver?

**Comment by ssahani** [comment]
closing infavour of https://github.com/systemd/systemd/pull/11716

**Pull request closed without merging**